ccmbioinfo/lfs
# CARDS-2514: New "locking" module, with basic locking feature

Initial implementation including:
- Locking restriction
- Lock node on subjects
  - Includes author who locked the node and time
  - Only locked subjects with a child lock node can be unlocked and only if their parent is unlocked
    to prevent partially unlocked child trees
- Basic UI for locking and unlocking
  - Single lock/unlock button based on current lock state

Testing:
- Verify that locking is opt-in:
  - Launch using `./start_cards.sh --test --dev` to make sure nothing has changed and that locking is disabled by default
  - Launch using `./start_cards.sh --test --dev --locking` to test that locking is can be enabled and is working as expected
  - Prepare and run docker with locking:
    - `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image --locking`
    - `docker-compose build && docker-compose up`
    - Test that locking is working as expected
    - `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
  - Prepare and run docker without locking: `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image`

Make sure that:
- Existing action buttons on the subject page work as before (delete, print preview)
- When a subject is locked, non-admin users cannot edit the subject or any of it's child forms
  - Currently any user can add or remove locks, this will be restricted by a future task
- Subjects cannot be locked or unlocked if they have a locked parent
- If you lock a child subject then lock a parent subject, unlocking the parent will not unlock the child
- If you lock a parent subject that has unlocked children (forms and/or subject), the child nodes will be locked too

https://github.com/data-team-uhn/cards/assets/20056751/b737e737-6a69-4e40-b732-7b0dda7e280b
